oak-blob-cloud/src/test/java/org/apache/jackrabbit/oak/blob/cloud/s3/TestS3DSWithSSEKMSwithKey.java

```diff
@@ -0,0 +1,109 @@
+        protected static long ONE_HUNDRED_MB = ONE_MB * 100;
+        protected static long ONE_GB = ONE_MB * ONE_MB;
+
+        @Override
+        @Before
+        public void setUp() throws Exception {
+            
+            super.setUp();
+            String randomKey = "2bf5110e-656b-4e22-98f8-1cc157d96447";
+            props.setProperty(S3Constants.S3_ENCRYPTION, S3Constants.S3_ENCRYPTION_SSE_KMS);
+            props.setProperty(S3Constants.S3_SSE_KMS_KEYID, randomKey);
+            props.setProperty("cacheSize", "0");
+        }
+
+        /**
+         * Test data migration enabling SSE_KMS encryption.
+         */
+        @Test
+        public void testDataMigration() {
+            try {
+                //manually close the setup ds and remove encryption
+                ds.close();
+                props.remove(S3Constants.S3_ENCRYPTION);
+                ds = createDataStore();
+
+                byte[] data = new byte[dataLength];
+                randomGen.nextBytes(data);
+                DataRecord rec = ds.addRecord(new ByteArrayInputStream(data));
+                Assert.assertEquals(data.length, rec.getLength());
+                assertRecord(data, rec);
+                ds.close();
+
+                // turn encryption now anc recreate datastore instance
+                props.setProperty(S3Constants.S3_ENCRYPTION, S3Constants.S3_ENCRYPTION_SSE_KMS);
+                props.setProperty(S3Constants.S3_RENAME_KEYS, "true");
+                ds = createDataStore();
+
+                rec = ds.getRecord(rec.getIdentifier());
+                Assert.assertEquals(data.length, rec.getLength());
+                assertRecord(data, rec);
+
+                randomGen.nextBytes(data);
+                ds.addRecord(new ByteArrayInputStream(data));
+                ds.close();
+            } catch (Exception e) {
+                LOG.error("error:", e);
+                fail(e.getMessage());
+            }
+        }
+
+        @Test
+        public void testInitiateDirectUploadUnlimitedURIs() throws DataRecordUploadException,
+                RepositoryException {
+            ConfigurableDataRecordAccessProvider ds
+                = (ConfigurableDataRecordAccessProvider) createDataStore();
+            long uploadSize = ONE_GB * 50;
+            int expectedNumURIs = 5000;
+            DataRecordUpload upload = ds.initiateDataRecordUpload(uploadSize, -1);
+            Assert.assertEquals(expectedNumURIs, upload.getUploadURIs().size());
+
+            uploadSize = ONE_GB * 100;
+            expectedNumURIs = 10000;
+            upload = ds.initiateDataRecordUpload(uploadSize, -1);
+            Assert.assertEquals(expectedNumURIs, upload.getUploadURIs().size());
+
+            uploadSize = ONE_GB * 200;
+            // expectedNumURIs still 10000, AWS limit
+            upload = ds.initiateDataRecordUpload(uploadSize, -1);
+            Assert.assertEquals(expectedNumURIs, upload.getUploadURIs().size());
+        }
+    }
```

Comment:
Missing test case for GeneratePreSignedUrlRequest and possible reason NPE for if condition not detected.